### PR TITLE
Improve Aspire logs search performance

### LINKED ISSUE #16981: Console log search is slow on larger log streams
## Summary

`aspire logs --search` appears to scale poorly as console log volume grows. In a small PR validation run with ~5,000 generated request logs, console-log queries took ~21-25 seconds even when the filtered output was tiny.

This was observed while validating #16939.

## Measurements

Generated **5,000 HTTP requests** to a C# Aspire `webapp`; each request emitted a structured log plus request telemetry.

| Command | Time |
|---|---:|
| `aspire logs webapp --search delta-search-marker --format Json` | **21.5s** |
| `aspire logs webapp --search perf-no-match-16939 --format Json` | **24.5s** |
| `aspire logs webapp --tail 100 --format Json` | **21.4s** |
| `aspire logs webapp --tail 100 --search delta-search-marker --format Json` | **22.5s** |
| `aspire otel logs webapp --limit 100 --format Json` | **367ms** |
| `aspire otel logs webapp --search delta-search-marker --limit 100 --format Json` | **393ms** |
| `aspire otel logs webapp --search perf-no-match-16939 --limit 100 --format Json` | **592ms** |
| `aspire otel traces webapp --search webapp --limit 100 --format Json` | **366ms** |
| `aspire otel spans webapp --search webapp --limit 100 --format Json` | **381ms** |

## Expected

`aspire logs --search` and `aspire logs --tail` should return quickly for larger console log streams, especially when the query returns few or no matching lines.

## Actual

The console log path took ~21-25 seconds for ~5,000 request-generated logs. `--tail 100` was about the same cost as searching all logs, and a no-match search returning almost no output still took ~25 seconds.

## Suspected cause

The current console log implementation appears to collect, parse, and sort the full console log snapshot before applying `--search` or `--tail`. That means filtering and tailing still pay the full snapshot-processing cost.

The OTEL telemetry API path looked fine at this scale because filtering happens in the dashboard telemetry API before returning the limited result set.

## Notes

This issue is for console logs only (`aspire logs`). The `aspire otel logs/spans/traces --search` path did not show the same behavior in this minimal test.

### PR DESCRIPTION
**Pull request by davidfowl**
## Description

Console log search is slow when an AppHost has large log streams because the CLI previously transferred every log line over JSON-RPC, parsed it, and then filtered locally. This changes the auxiliary backchannel so newer AppHosts can apply resource, search, hidden-resource, and single-resource tail filters before streaming logs back to the CLI. The CLI keeps its final client-side filtering for compatibility with older AppHosts and for parsed/display-name semantics.

This also adds profiling-only JSON-RPC spans for the CLI and Hosting backchannel path, with explicit trace context propagation in request objects so `aspire logs --search` can be measured end to end across processes. Newer AppHosts also expose an aux.v3 batched console-log stream so the CLI can reduce per-line JSON-RPC stream overhead without changing the existing line-based contract.

### User-facing usage

Users keep using the same `aspire logs` command. The optimization is automatic when the connected AppHost supports the v2/v3 auxiliary backchannel:

```bash
aspire logs producer --search needle --tail 75 --format Json
```

### Performance

E2E harness: 75,000 emitted log lines, 75 matching `needle` lines.

| Path | Wall clock | JSON-RPC stream items | Hosting log span |
| --- | ---: | ---: | ---: |
| Before: legacy client-side filtering | 131.71s | 75,002 lines | ~131s |
| After: server-side search filtering | 0.90s | 75 lines | ~255ms |
| After: server-side filtering + batched streaming | 0.92s avg (0.87s-0.95s) | 1 batch | not re-captured |

Batching drops the filtered tail result from 75 JSON-RPC stream frames to 1 batch in this scenario. The wall clock is comparable to the line-streaming server-side filter path because command startup and AppHost log scanning dominate after the full 75,000-line transfer is removed.

### Validation

- `dotnet build tests/Aspire.Cli.Tests/Aspire.Cli.Tests.csproj --no-restore /p:SkipNativeBuild=true`
- `dotnet build tests/Aspire.Hosting.Tests/Aspire.Hosting.Tests.csproj --no-restore /p:SkipNativeBuild=true`
- `dotnet test --project tests/Aspire.Cli.Tests/Aspire.Cli.Tests.csproj --no-build --no-launch-profile -- --filter-class "*.LogsCommandTests" --filter-not-trait "quarantined=true" --filter-not-trait "outerloop=true"`
- `dotnet test --project tests/Aspire.Hosting.Tests/Aspire.Hosting.Tests.csproj --no-build --no-launch-profile -- --filter-method "*.GetConsoleLogsAsync_AppliesSearchAndTailForSingleResource" --filter-method "*.GetConsoleLogBatchesAsync_AppliesSearchAndTailForSingleResource" --filter-method "*.GetConsoleLogsAsync_DoesNotApplyTailAcrossMultipleResources" --filter-method "*.GetConsoleLogsAsync_ExcludesHiddenResourcesWhenStreamingAllResources" --filter-not-trait "quarantined=true" --filter-not-trait "outerloop=true"`
- `dotnet test --project tests/Aspire.Hosting.Tests/Aspire.Hosting.Tests.csproj --no-build --no-launch-profile -- --filter-method "*.GetCapabilitiesAsyncReturnsCurrentCapabilities" --filter-not-trait "quarantined=true" --filter-not-trait "outerloop=true"`
- `dotnet test --project tests/Aspire.Hosting.Tests/Aspire.Hosting.Tests.csproj --no-build --no-launch-profile -- --filter-method "*.BackchannelTypes_FollowContractRules" --filter-not-trait "quarantined=true" --filter-not-trait "outerloop=true"`
- `./eng/scripts/verify-startup-otel.sh --skip-build --target-aspire-path artifacts/bin/Aspire.Cli/Debug/net10.0/aspire --profiler-aspire-path artifacts/bin/Aspire.Cli/Debug/net10.0/aspire --post-start-delay 1`
- Manual E2E `aspire logs producer --search needle --tail 75 --format Json` against a detached AppHost returned 75 matching lines from `line 001000 needle` through `line 075000 needle`; repeated batched runs measured `0.93s`, `0.95s`, and `0.87s` wall clock.

Fixes #16981

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No

### CODE REVIEW

File: src/Aspire.Cli/Commands/LogsCommand.cs [hunk 1]
```diff
@@ -327,7 +329,15 @@ private async Task<int> ExecuteWatchAsync(
         }
 
         // Now stream new logs
-        await foreach (var logLine in connection.GetResourceLogsAsync(resourceName, follow: true, cancellationToken).ConfigureAwait(false))
+        var followRequest = new GetConsoleLogsRequest
+        {
+            ResourceName = resourceName,
+            Follow = true,
+            Search = search,
+            IncludeHidden = resourceName is not null || resourceWatcher.IncludeHidden
```
Comment (JamesNK) [inline src/Aspire.Cli/Commands/LogsCommand.cs:338]: Minor: the `followRequest` correctly omits `Tail` (tail only applies to the initial snapshot), but `GetConsoleLogsRequest` exposes both `Follow` and `Tail` with no constraint preventing `Follow = true, Tail = 50`. The server silently ignores `Tail` when `follow` is true because `serverTailLineCount` is 0 for follow mode. If a future caller passes both, the `Tail` is silently dropped. Consider adding a comment (or `Debug.Assert`) on the server side in the `serverTailLineCount` calculation to document that `Tail` is intentionally ignored for follow streams.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs [hunk 2]
```diff
@@ -234,6 +264,7 @@ public async Task<CallMcpToolResponse> CallMcpToolAsync(CallMcpToolRequest reque
     /// <returns>The stop response.</returns>
     public async Task<StopAppHostResponse> StopAsync(StopAppHostRequest? request = null, CancellationToken cancellationToken = default)
     {
+        using var activity = profilingTelemetry.StartJsonRpcServerCall(nameof(StopAsync), streaming: false, request?.ProfilingContext);
         _ = request; // Exit code not yet used, but available for future expansion
```
Comment (JamesNK) [inline src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs]: Stale dead-code discard: `_ = request;` is now dead code because the line above already accesses `request?.ProfilingContext`. The comment "Exit code not yet used, but available for future expansion" is misleading since the request is now consumed for profiling context. Consider removing the discard or updating the comment to only reference `ExitCode`.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs [hunk 3]
```diff
@@ -918,78 +944,197 @@ public async IAsyncEnumerable<ResourceLogLine> GetResourceLogsAsync(
...
-        if (resourcesToLog.Count == 0)
-        {
-            yield break;
-        }
+            if (!includeHidden && resourceName is null)
+            {
+                resourcesToLog = resourcesToLog
+                    .Where(resolvedResourceName => !IsHiddenResource(notificationService, resolvedResourceName))
+                    .ToList();
+            }
```
Comment (JamesNK) [inline src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs:1043]: The hidden-resource filter only applies when `resourceName is null`: `if (!includeHidden && resourceName is null)`. This means a caller that passes a specific `resourceName` with `IncludeHidden = false` will still get logs for that resource even if it's hidden.

The CLI always sets `IncludeHidden = true` when `resourceName is not null` (see `LogsCommand.cs`), so this is benign today. But the server contract allows the combination and silently ignores it. If this is intentional (i.e., requesting a resource by name always grants access), a brief comment would clarify the invariant and prevent future callers from relying on `IncludeHidden = false` with a named resource.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs [hunk 4]
```diff
@@ -918,78 +944,197 @@ public async IAsyncEnumerable<ResourceLogLine> GetResourceLogsAsync(
...
+
+            // Only apply tailing on the server when a single resource is requested. For multiple
+            // resources, the CLI needs the parsed/merged ordering across resources before it can
+            // choose the final tail entries without changing observable output order.
+            var serverTailLineCount = !follow && tail.GetValueOrDefault() > 0 && resourcesToLog.Count == 1
+                ? tail.GetValueOrDefault()
+                : 0;
+            Queue<ResourceLogLine>? tailBuffer = serverTailLineCount > 0 ? new Queue<ResourceLogLine>(serverTailLineCount) : null;
+
```
Comment (JamesNK) [inline src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs:1067]: `new Queue<ResourceLogLine>(serverTailLineCount)` sets initial capacity, not maximum size. The sliding-window enforcement via `Dequeue`/`Enqueue` below is correct, so this isn't a bug. But the capacity constructor parameter can mislead readers into thinking the queue is bounded. Consider adding a brief comment noting that capacity is a sizing hint and the max is enforced by the dequeue logic below.

**Review by JamesNK — COMMENTED** [review]
4 items flagged: 1 stale dead-code discard, 1 hidden-resource filter contract clarity, 1 follow/tail asymmetry documentation, 1 queue capacity clarification. All are non-blocking observations — no bugs found.

File: src/Aspire.Cli/Backchannel/ProfilingJsonRpcExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        }
+        catch (Exception ex)
+        {
+            activity.SetError(ex);
+            activity.Dispose();
+            throw;
+        }
+    }
+
+    private static async IAsyncEnumerable<T> EnumerateWithProfiling<T>(
```
Comment (JamesNK) [inline src/Aspire.Cli/Backchannel/ProfilingJsonRpcExtensions.cs:101]: `InvokeStreamingWithProfilingAsync` intentionally does not `using`-dispose the activity, delegating ownership to `EnumerateWithProfiling` which disposes it in `finally`. If the returned `IAsyncEnumerable<T>` is obtained but never enumerated (caller never calls `GetAsyncEnumerator`), the activity leaks. All current call sites immediately enumerate so this is theoretical-only, but the ownership transfer is non-obvious — a brief comment on `InvokeStreamingWithProfilingAsync` noting "activity ownership transfers to the returned enumerable; callers must enumerate or dispose will not occur" would help future maintainers.

**Review by JamesNK — COMMENTED** [review]
Additional finding: activity ownership transfer in streaming profiling helper.

**Comment by JamesNK** [comment]
> The CLI keeps its final client-side filtering for compatibility with older AppHosts and for parsed/display-name semantics.

Search doesn't need to be kept in the client. It's new in 13.3.

**Comment by JamesNK** [comment]
I'm surprised how slow this is without filtering in the app host. All other times the perf bottle neck has been DCP streaming to the app host.

Is JSON-RPC really bad at streaming?

**Comment by davidfowl** [comment]
Thanks for the overview. No code change was needed for this generated summary; I addressed the actionable review threads separately and pushed the follow-up commit.

**Comment by davidfowl** [comment]
Accepted these observations. I removed the stale discard and added comments clarifying hidden named-resource requests, follow/tail behavior, and the fixed-size tail queue window.

**Comment by davidfowl** [comment]
Accepted. I added a comment at the streaming profiling helper to document that activity ownership transfers to the returned enumerable and disposal depends on enumeration.

File: src/Aspire.Cli/Commands/LogsCommand.cs [hunk 6]
```diff
@@ -364,13 +374,27 @@ private static async Task<IList<LogEntry>> CollectLogsAsync(
...
+        var request = new GetConsoleLogsRequest
+        {
+            ResourceName = resourceName,
+            Follow = false,
+            Search = search,
+            Tail = tail,
+            IncludeHidden = resourceName is not null || resourceWatcher.IncludeHidden
+        };
+
+        await foreach (var logLine in connection.GetConsoleLogsAsync(request, cancellationToken).ConfigureAwait(false))
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Cli/Commands/LogsCommand.cs]: Accepted. All-resource snapshot requests now stay on the legacy GetResourceLogsAsync RPC so older aux.v2 AppHosts are not sent GetConsoleLogsRequest with a null ResourceName; the CLI still applies search and tail locally.

File: src/Aspire.Cli/Commands/LogsCommand.cs [hunk 7]
```diff
@@ -327,7 +329,15 @@ private async Task<int> ExecuteWatchAsync(
...
-        await foreach (var logLine in connection.GetResourceLogsAsync(resourceName, follow: true, cancellationToken).ConfigureAwait(false))
+        var followRequest = new GetConsoleLogsRequest
+        {
+            ResourceName = resourceName,
+            Follow = true,
+            Search = search,
+            IncludeHidden = resourceName is not null || resourceWatcher.IncludeHidden
+        };
+
+        await foreach (var logLine in connection.GetConsoleLogsAsync(followRequest, cancellationToken).ConfigureAwait(false))
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Cli/Commands/LogsCommand.cs]: Accepted. All-resource follow requests now also use the legacy GetResourceLogsAsync RPC, avoiding the nullable ResourceName shape against older aux.v2 AppHosts while keeping client-side filtering.

File: src/Aspire.Cli/Commands/LogsCommand.cs (hunk 1 above)
Comment (davidfowl) [reply src/Aspire.Cli/Commands/LogsCommand.cs:338]: Accepted. I added server-side comments explaining that Tail is intentionally ignored for follow streams, and that multi-resource tail remains client-owned to preserve merged ordering.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs]: Accepted. I removed the stale discard and updated the parameter docs so they no longer imply the whole request is unused.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs (hunk 3 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs:1043]: Accepted. I added a comment documenting that IncludeHidden filters only all-resource streams, while a named resource request is treated as an explicit request even if the resource is hidden.

File: src/Aspire.Cli/Backchannel/ProfilingJsonRpcExtensions.cs (hunk 5 above)
Comment (davidfowl) [reply src/Aspire.Cli/Backchannel/ProfilingJsonRpcExtensions.cs:101]: Accepted. I added a comment documenting the streaming activity ownership transfer and the caveat that a returned enumerable must be enumerated for the profiling activity to be disposed.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs [hunk 8]
```diff
@@ -170,34 +236,67 @@ internal sealed class GetResourcesResponse
...
+    /// </summary>
+    public int? Tail { get; init; }
+
+    /// <summary>
+    /// Gets whether hidden resources should be included when no resource name is specified.
+    /// </summary>
+    public bool IncludeHidden { get; init; }
+
+    /// <inheritdoc />
+    public override GetConsoleLogsRequest WithProfilingContext(BackchannelProfilingContext profilingContext) => new()
```
Comment (IEvangelist) [inline src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Maintainability concern: each BackchannelRequest subclass overrides WithProfilingContext and hand-copies every property. If a future contributor adds a new property to (e.g.) GetConsoleLogsRequest and forgets to update this override, the new property is silently sent as default over the wire — but only when profiling is enabled, so tests run with profiling off won't catch it. The AOT/trimming rationale on the base class is valid, but consider adding a reflection-based unit test to BackchannelContractTests that asserts 
equest.WithProfilingContext(pc) round-trips every public init property (the test runs outside AOT and won't affect runtime trimming).

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs [hunk 9]
```diff
@@ -919,77 +964,241 @@ public async IAsyncEnumerable<ResourceLogLine> GetResourceLogsAsync(
...
+            activity.SetJsonRpcStreamItemCount(yieldedCount);
         }
     }
 
+    private static bool IsHiddenResource(ResourceNotificationService notificationService, string resourceName)
+    {
+        return notificationService.TryGetCurrentState(resourceName, out var resourceEvent) && resourceEvent.Snapshot.IsHidden;
+    }
+
+    private static bool MatchesSearch(ResourceLogLine logLine, string? search)
```
Comment (IEvangelist) [inline src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs:1184]: Server-side filter runs on the raw logLine.Content, which may contain ANSI escape sequences and timestamp prefixes. The CLI comment in LogsCommand.cs (preserves the CLI's parsed-log search semantics for any edge cases the server-side raw search misses) only covers the over-match direction (server matches because of timestamp, CLI filters out). The opposite direction is a real regression: if log content contains an ANSI escape mid-token (e.g., `Read\x1b[31my needle`), a search for `Ready` won't match the raw bytes here, the line is dropped, and the CLI never sees it. Before this PR, the CLI parsed/stripped ANSI first and matched. Either document the trade-off honestly in the comment, or normalize ANSI escapes on the server before substring matching.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs [hunk 10]
```diff
@@ -170,34 +236,67 @@ internal sealed class GetResourcesResponse
...
 /// Request for getting console logs.
 /// </summary>
-internal sealed class GetConsoleLogsRequest
+internal sealed class GetConsoleLogsRequest : BackchannelRequest
 {
     /// <summary>
     /// Gets the resource name to get logs for.
     /// </summary>
-    public required string ResourceName { get; init; }
+    public string? ResourceName { get; init; }
```
Comment (IEvangelist) [inline src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs:254]: ResourceName was 
equired string in V2 and is now string?. The spec at docs/specs/cli-backchannel.md still states rule 4. Required properties use 'required' keyword and doesn't mention the relaxation. The CLI handles compatibility by routing all-resource queries through legacy GetResourceLogsAsync, but a third-party CLI talking to a new AppHost over V2 can now legitimately send a null ResourceName for all-resource logs. Worth a note in the spec under V2/V3 stating that ResourceName is now optional in V3, and that V2 callers should keep using GetResourceLogsAsync for all-resource queries.

**Review by IEvangelist — APPROVED** [review]
Approving. Three non-blocking observations from review:

1. WithProfilingContext hand-copies properties in 11 request types — adding a new property without updating the override silently drops it on the wire when profiling is enabled. A reflection-based contract test would prevent this.
2. Server-side raw-content MatchesSearch can drop lines that the CLI's parsed-log search would match (ANSI-escape edge case).
3. GetConsoleLogsRequest.ResourceName relaxed from 
equired to string? is not reflected in the spec doc.

Nice perf win — 131s → 0.9s is a great result.

**Comment by davidfowl** [comment]
Accepted all three observations. I added the WithProfilingContext property-preservation contract test, normalized ANSI stripping in the server and CLI search paths with regression coverage, and documented ResourceName V2/V3 compatibility in the backchannel spec.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs (hunk 8 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Accepted. I added a reflection-based BackchannelContractTests case that populates each request type, calls WithProfilingContext, and verifies every public init property is preserved while ProfilingContext is replaced.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs (hunk 9 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs:1184]: Accepted. ResourceLoggerService stores raw ANSI content, so I reused AnsiParser.StripControlSequences in the server pre-filter and CLI final search path and added split-token regression coverage.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs (hunk 10 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs:254]: Accepted. I added a console log request compatibility note to the backchannel spec explaining ResourceName is optional for v3 and that V2 all-resource callers should keep using GetResourceLogsAsync.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs [hunk 11]
```diff
@@ -61,10 +66,56 @@ internal static class KnownCommandVisibility
...
+
+/// <summary>
+/// Base class for auxiliary backchannel request-object RPC parameters.
+/// </summary>
+internal abstract class BackchannelRequest
+{
+    /// <summary>
+    /// Gets profiling trace context propagated by the CLI.
+    /// </summary>
+    public BackchannelProfilingContext? ProfilingContext { get; init; }
```
Comment (JamesNK) [inline src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: This is sending tracing information in the JSON body. Why not use OTEL standardized HTTP traceparent, tracestate headers?

Comment (JamesNK) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Is it because we're using UDS/named pipes instead of HTTP?

Comment (JamesNK) [inline src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: I'd prefer that name wasn't tied to profiling. This is trace information that should be generic to any tracing.

What about instead:

```
public BackchannelTraceContext? TraceContext { get; init; }
```

And rename other types to match.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs [hunk 12]
```diff
@@ -61,10 +66,56 @@ internal static class KnownCommandVisibility
...
 
+/// <summary>
+/// Profiling trace context propagated over the auxiliary backchannel.
+/// </summary>
+internal sealed class BackchannelProfilingContext
+{
+    /// <summary>
+    /// Gets the profiling session ID.
+    /// </summary>
+    public string? ProfilingSessionId { get; init; }
```
Comment (JamesNK) [inline src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: This is passed as baggage elseware. What about here? Could this instead be:

```
public Dictionary<string, string> Baggage { get; } = [];
```

And then the session ID is just data in the baggage.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs (hunk 11 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: It's not HTTP it's JSON RPC over UDS (like TCP). It's like adding it to signalr (it needs to be part of the protocol).

File: src/Aspire.Hosting/Diagnostics/ProfilingTelemetry.cs [hunk 13]
```diff
@@ -339,6 +383,23 @@ private static void AddProfilingSessionId(Activity? activity, IConfiguration? co
...
+        {
+            return;
+        }
+
+        var sessionId = !string.IsNullOrEmpty(profilingContext?.ProfilingSessionId)
+            ? profilingContext.ProfilingSessionId
+            : GetConfigurationValue(configuration, KnownConfigNames.ProfilingSessionId, KnownConfigNames.Legacy.StartupOperationId);
+        if (!string.IsNullOrEmpty(sessionId))
+        {
+            activity.SetTag(Tags.ProfilingSessionId, sessionId);
```
Comment (JamesNK) [inline src/Aspire.Hosting/Diagnostics/ProfilingTelemetry.cs:404]: Should this be baggage and not a tag? So it's propagated in external calls the app host makes.

**Comment by JamesNK** [comment]
I don't know if this was an existing issue with `--search` or not, but from testing:

Search with empty string returns everything:
```
aspire logs --search "" --timestamps --tail 2
```
This makes sense.

Search with no value returns nothing:
```
aspire logs --search --timestamps --tail 2
```
I feel like it should either return everything, or CLI option validation should complain that a value is required.

Edit: I investigated. The value of search is always the value after it, even if it matches another option, i.e. `--timestamps`. Not a bug.

**Comment by JamesNK** [comment]
Performance oddity:
1. In the stress playground app I ran the log message limit command on `stress-apiservice`
2. With search it is very fast (sub 1 second): `aspire logs --search "dashboard" --timestamps --tail 2`
3. WIthout search, and only tail, it is much slower (20 seconds): `aspire logs --timestamps --tail 2`

Is tail filtered in the CLI?

Edit: I investigated and tail isn't filtered in the app host when there are multiple resources. Is this something we could improve? Most of the time there won't be a single resource specified.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs [hunk 14]
```diff
@@ -919,75 +965,254 @@ public async IAsyncEnumerable<ResourceLogLine> GetResourceLogsAsync(
...
-            logger.LogWarning("Resource '{ResourceName}' not found. No logs will be returned.", resourceName);
-            yield break;
+            yield return logLine;
         }
+    }
 
-        if (resourcesToLog.Count == 0)
-        {
-            yield break;
+    private async IAsyncEnumerable<ResourceLogBatch> GetResourceLogBatchesAsync(
```
Comment (JamesNK) [inline src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs:982]: There is a ChannelExtensions type that has a helper method for batching results from a channel. Could do the same thing with IAsyncEnumerable to make this logic reusable elsewhere. I suspect we'll batch more in backchannels in the future.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs (hunk 11 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Partially accepted. This transport does not have HTTP headers, but I changed this to use StreamJsonRpc's ActivityTracingStrategy so W3C traceparent/tracestate travel as JSON-RPC request envelope metadata instead of Aspire request-body fields.

Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Yes, exactly. Since this is JSON-RPC over UDS/named pipes, the protocol-level equivalent is StreamJsonRpc's W3C trace context fields on the JSON-RPC request envelope, and the PR now uses that.

Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Accepted. I renamed the request metadata to TraceContext/BackchannelTraceContext/WithTraceContext so it is not tied to profiling, with profiling-specific meaning limited to the baggage/tag values.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs (hunk 12 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Accepted. The request metadata now carries a Baggage dictionary, and the profiling session ID flows as aspire.profiling.session_id baggage instead of a dedicated ProfilingSessionId property.

File: src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs (hunk 11 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/BackchannelDataTypes.cs]: Agreed. I changed the implementation to treat traceparent/tracestate as JSON-RPC protocol metadata via StreamJsonRpc rather than request payload fields.

File: src/Aspire.Hosting/Diagnostics/ProfilingTelemetry.cs (hunk 13 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Diagnostics/ProfilingTelemetry.cs:404]: Accepted. Hosting now restores the profiling session ID as Activity baggage as well as tags, so downstream calls can propagate it while existing tag-based queries still work.

File: src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs (hunk 14 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Backchannel/AuxiliaryBackchannelRpcTarget.cs:982]: Accepted. I added a reusable `IAsyncEnumerable<T>.GetBatchesAsync` helper next to the existing channel batching helper and switched the console log batch paths to use it.
